pkg/services/auth/auth.go

```diff
@@ -0,0 +1,242 @@
+func NewAuthService(fluxClient flux.Flux, k8sClient client.Client, providerName gitproviders.GitProviderName, l logger.Logger, token string) (*AuthService, error) {
+	provider, err := gitproviders.New(gitproviders.Config{Provider: providerName, Token: token})
+	if err != nil {
+		return nil, fmt.Errorf("error creating git provider client: %w", err)
+	}
+	return &AuthService{
+		logger:      l,
+		fluxClient:  fluxClient,
+		k8sClient:   k8sClient,
+		gitProvider: provider,
+		// We do not have a gitClient yet. We need to get or create deploy keys first.
+		// This will get populated by the SetupDeployKeys call.
+		// Listing it here for the sake of readability (it would still be nil without this entry).
+		gitClient: nil,
+	}, nil
+}
+
+// SetupDeployKey instruments the AuthService with the neccesary data to provide a git client that uses deploy keys.
+// The user MUST call this function before calling .GitClient(), else an error will be returned.
+// The work of setting up deploy keys was NOT done in the constructor for the sake being explicit about what is happening.
+func (a *AuthService) SetupDeployKey(ctx context.Context, name types.NamespacedName, targetName string, repo gitproviders.NormalizedRepoURL) error {
+	owner := repo.Owner()
+	repoName := repo.RepositoryName()
+	accountType, err := a.gitProvider.GetAccountType(repo.Owner())
+	if err != nil {
+		return fmt.Errorf("error getting account type: %w", err)
+	}
+
+	repoInfo, err := a.gitProvider.GetRepoInfo(accountType, repo.Owner(), repo.RepositoryName())
+	if err != nil {
+		return fmt.Errorf("error getting repo info: %w", err)
+	}
+
+	if repoInfo.Visibility != nil && *repoInfo.Visibility == gitprovider.RepositoryVisibilityPublic {
+		// This is a public repo. We don't need to add deploy keys to it.
+		a.gitClient = git.New(nil)
+		return nil
+	}
+
+	deployKeyExists, err := a.gitProvider.DeployKeyExists(owner, repoName)
+	if err != nil {
+		return fmt.Errorf("failed check for existing deploy key: %w", err)
+	}
+
+	if deployKeyExists {
+		a.logger.Println("Existing deploy key found")
+		secretName := types.NamespacedName{
+			Name:      gitproviders.CreateAppSecretName(targetName, repo),
+			Namespace: name.Namespace,
+		}
+		// The deploy key was found on the Git Provider, fetch it from the cluster.
+		secret, err := a.retreiveDeployKey(ctx, secretName)
+		if apierrors.IsNotFound(err) {
+			// Edge case where the deploy key exists on the Git Provider, but not on the cluster.
+			// Users might end up here if we uploaded the deploy key, but it failed to save on the cluster.
+			// TODO: What should we do to help the user out here? We can't read the key data from the Git Provider,
+			// do we delete and recreate?
+			return fmt.Errorf("deploy key does not exist on cluster: %w", err)
+		} else if err != nil {
+			return fmt.Errorf("error retrieving deploy key: %w", err)
+		}
+
+		b := extractPrivateKey(secret)
+
+		pubKey, err := makePublicKey(b)
```

Comment:
Note: if you fix the above shadow variable then this will become a new shadow variable, so I would rename this one as well.